### DISCUSSION TITLE
[General] hardhat-zksync-verify appears to overwrite a hardhat task

### DISCUSSION THREAD

**Discussion by jordaniza** [zkSync-Community-Hub/zksync-developers#164] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

Trying to run HH Tests for ZkSync and not able to run yarn hardhat test. 

### What is the specific issue or error you're encountering?

Error HH210: Redefinition of task verify failed. Unsupported operation adding positional param definitions in an overridden task.

### Can you share the error messages or logs you're receiving, if any?

Error HH210: Redefinition of task verify failed. Unsupported operation adding positional param definitions in an overridden task.

### Have you made any recent changes to the contract before encountering this issue?

Updated packages to latest versions

### Are there any external libraries or contracts that your contract interacts with?

```json
  "devDependencies": {
    "@ethersproject/abi": "^5.7.0",
    "@ethersproject/providers": "^5.7.2",
    "@matterlabs/hardhat-zksync-deploy": "latest",
    "@matterlabs/hardhat-zksync-solc": "latest",
    "@matterlabs/hardhat-zksync-upgradable": "latest",
    "@matterlabs/hardhat-zksync-verify": "^0.2.0",
    "@nomicfoundation/hardhat-chai-matchers": "^1.0.6",
    "@nomicfoundation/hardhat-network-helpers": "^1.0.8",
    "@nomicfoundation/hardhat-toolbox": "^2.0.1",
    "@nomicfoundation/hardhat-verify": "^1.1.1",
    "@nomiclabs/hardhat-ethers": "^2.2.2",
    "@nomiclabs/hardhat-etherscan": "^3.1.7",
    "@openzeppelin/contracts": "^4.9.3",
    "@openzeppelin/contracts-upgradeable": "^4.9.3",
    "@openzeppelin/hardhat-upgrades": "^1.22.1",
    "@typechain/ethers-v5": "^10.2.0",
    "@typechain/hardhat": "^6.1.5",
    "@types/chai": "^4.3.4",
    "@types/mocha": "^10.0.1",
    "chai": "^4.3.7",
    "chai-as-promised": "^7.1.1",
    "dotenv": "^16.0.3",
    "ethers": "^5.7.2",
    "hardhat": "^2.12.4",
    "hardhat-gas-reporter": "^1.0.9",
    "mocha": "^10.2.0",
    "solidity-coverage": "^0.8.2",
    "ts-node": "^10.9.1",
    "typechain": "^8.3.1",
    "typescript": "^5.2.2",
    "zksync-web3": "latest"
  },
```

### Can you provide the relevant portions of your contract code where the issue is occurring?

The below code swaps between the HH EVM and ZK EVM depending on env vars, to allow for faster HH tests when testing base contract logic. 

```ts
import { HardhatUserConfig } from "hardhat/config";
import "@nomicfoundation/hardhat-toolbox";
import "@typechain/hardhat";
import "solidity-coverage";
import "@openzeppelin/hardhat-upgrades";
import dotenv from "dotenv";

import "@matterlabs/hardhat-zksync-upgradable";
import "@matterlabs/hardhat-zksync-deploy";
import "@matterlabs/hardhat-zksync-solc";
// bugged as of now
import "@matterlabs/hardhat-zksync-verify";

dotenv.config();

// Environment variables
const useZkEVM = process.env.USE_ZKEVM === "true";
const isTestEnv = process.env.NODE_ENV == "test";

const zkSyncTestnet = isTestEnv
  ? {
      url: "http://localhost:3050",
      ethNetwork: "http://localhost:8545",
      zksync: true,
    }
  : {
      url: "https://zksync2-testnet.zksync.dev",
      ethNetwork: "goerli",
      zksync: true,
      // verifyURL: "https://zksync2-testnet-explorer.zksync.dev/contract_verification",
    };

const commonConfig: HardhatUserConfig = {
  solidity: {
    version: "0.8.17",
  },
  typechain: {
    target: "ethers-v5",
  },
};

const zkConfig: HardhatUserConfig = {
  ...commonConfig,
  zksolc: {
    version: "latest",
    settings: {},
  },
  defaultNetwork: "zkSyncTestnet",
  networks: {
    hardhat: {
      zksync: true,
    },
    zkSyncTestnet,
  },
  paths: {
    tests: "./test/zk",
  },
};

const defaultConfig: HardhatUserConfig = {
  ...commonConfig,
  paths: {
    tests: "./test/hh",
  },
};

export default useZkEVM ? zkConfig : defaultConfig;
```

### Have you tried to isolate the problem, and if so, what were the results?

This error occurs if the "@matterlabs/hardhat-zksync-verify" is imported into a hardhat config. Removing this line resolves the problem, but that then prevents verification of ZKBytecode

### What steps have you already taken to try to resolve the issue?

I've tried rescaffolding a fresh package with the CLI and I get the same issue.

**Comment by AlbionaHoti** [comment]
Hi @jordaniza,

Thanks for opening this discussion! I ran into the same one today; the solution to this is to place the `import "@matterlabs/hardhat-zksync-verify";` before the `import "@matterlabs/hardhat-zksync-upgradable";
`. See if it works, and if not, please reach out again here. I'll wait a few days for your response; otherwise, I will mark this as answered.

**Reply by jordaniza** [reply]
Just tried a few different reorderings @AlbionaHoti, none seem to work unfortunately.

**Reply by AlbionaHoti** [reply]
Is it possible for us to get a repo for the code so we can run it locally?

**Reply by jordaniza** [reply]
Sure I can invite you to the repo if you like

**Reply by idea404** [reply]
To provide some background on this error:

1. A few months back, the Hardhat team introduced a new package named `hardhat-verify`. This marked the beginning of their transition from the `hardhat-etherscan` package to this new one. As a consequence, several packages, including `hardhat-toolbox`, began adopting this new package.
2. At the same time, our team released an updated version of `hardhat-zksync-verify` (version 0.2.0, which introduced breaking changes). This version relies on `hardhat-verify`, suggesting that we're either modifying or extending tasks from this new package.
3. If we delve into the Hardhat configuration file shared above, we see that an older iteration of `hardhat-toolbox` (version 2.0.1) is being imported initially. This action registers all tasks associated with the older `hardhat-etherscan` plugin.
4. The issue crops up when our plugin is subsequently imported. Due to the tasks from the older `hardhat-etherscan` plugin being already registered, our plugin's import leads to unintended task overrides, causing the observed bug.

**Reply by jordaniza** [reply]
Hey @AlbionaHoti, let me know if you still need repo access but from the discord seems like I'm not the only one and @idea404 seems to have articulated the error correctly. 

Let me know if the team would like assistance on modifying the zksync verify package to be compatible with hardhat toolbox, or if this is already in hand - happy to support if needed or just be patient.

**Reply by idea404** [reply, marked as answer]
@jordaniza would be great if you could open a PR to address this!

I opened [this issue](https://github.com/matter-labs/hardhat-zksync/issues/404) on the repo which you can mention in your PR.

You can add @mpopovac-txfusion as a reviewer once you open it.

**Reply by jordaniza** [reply]
cool I'll take a look when I'm back working on the ZkSync stuff

**Comment by steven2308** [comment]
I'm having the same issue. Here are my dependencies:

```
"devDependencies": {
  "@matterlabs/hardhat-zksync-deploy": "^0.6.5",
  "@matterlabs/hardhat-zksync-solc": "^0.4.2",
  "@matterlabs/hardhat-zksync-verify": "^0.2.0",
  "@ethersproject/abi": "^5.4.7",
  "@ethersproject/providers": "^5.4.7",
  "@nomicfoundation/hardhat-verify": "^1.1.0",
  "@nomicfoundation/hardhat-chai-matchers": "^1.0.6",
  "@nomicfoundation/hardhat-network-helpers": "^1.0.8",
  "@nomicfoundation/hardhat-toolbox": "^2.0.2",
  "@nomiclabs/hardhat-ethers": "^2.2.2",
  "@nomiclabs/hardhat-etherscan": "^3.1.7",
  "@openzeppelin/test-helpers": "^0.5.16",
  "@typechain/ethers-v5": "^10.2.0",
  "@typechain/hardhat": "^6.1.2",
  "@types/chai": "^4.3.4",
  "@types/mocha": "^10.0.1",
  "@types/node": "^18.11.18",
  "@types/rimraf": "^4.0.5",
  "chai": "^4.3.7",
  "dotenv": "^16.0.3",
  "ethers": "^5.7.2",
  "hardhat": "^2.17.0",
  "hardhat-contract-sizer": "^2.10.0",
  "hardhat-gas-reporter": "^1.0.9",
  "prettier": "2.8.3",
  "prettier-plugin-solidity": "^1.1.3",
  "solc": "^0.8.21",
  "solhint": "^3.4.0",
  "solidity-coverage": "^0.8.4",
  "ts-node": "^10.9.1",
  "typechain": "^8.1.1",
  "typescript": "^4.9.4",
  "walk-sync": "^3.0.0",
  "zksync-web3": "^0.14.3"
},
```

And the imports on my hardhat config:
```
import * as dotenv from 'dotenv';
import '@nomicfoundation/hardhat-toolbox';
import '@typechain/hardhat';
import 'hardhat-contract-sizer';
import 'hardhat-gas-reporter';
import 'solidity-coverage';
import '@matterlabs/hardhat-zksync-verify';
import '@matterlabs/hardhat-zksync-deploy';
import '@matterlabs/hardhat-zksync-solc';
```

I tried changing the order and commenting imports for typechain, contract sizer, gas reporter and coverage, but but it did not help.
```
Error HH210: Redefinition of task verify failed. Unsupported operation adding positional param definitions in an overridden task.
```

**Comment by patelkaren** [comment]
The error HH210: Redefinition of task verify failed is likely caused by a conflict between the verify task defined in the @nomiclabs/hardhat-etherscan plugin and the one in the @matterlabs/hardhat-zksync-verify plugin. You can try removing one of these plugins from your configuration or renaming the verify task in one of the plugins to resolve the conflict.

**Comment by kaaron9** [comment]
This error occurs due to a redefinition conflict in the hardhat-zksync-verify plugin when using positional param definitions in an overridden task. Removing the plugin resolves the issue, but it prevents verification of ZKBytecode. Consider reaching out to the plugin maintainers for a resolution or explore alternative verification methods.

**Comment by golfdutoresste** [comment]
You can try removing the "@matterlabs/hardhat-zksync-verify" import from your hardhat config as a temporary solution.
This will resolve the HH210 error, although it will prevent the verification of ZKBytecode.
 Keep an eye on updates to the ZkSync plugin to see if this issue gets resolved in future versions.

**Comment by hollinsmic** [comment]
The user is trying to run HH tests for ZkSync on Testnet but is encountering an error message "Error HH210: Redefinition of task verify failed. Unsupported operation adding positional param definitions in an overridden task." when trying to run yarn hardhat test. They have updated packages to the latest versions but have not mentioned any external libraries or contracts that their contract interacts with.
